pkg/controller/statefulset/stateful_set_control.go

```diff
@@ -759,7 +759,7 @@ func updateStatefulSetAfterInvariantEstablished(
 	// those pods and count the successful deletions. Update the status with the correct number of deletions.
 	unavailablePods := 0
 	for target := len(replicas) - 1; target >= 0; target-- {
-		if !isHealthy(replicas[target]) {
+		if !isRunningAndAvailable(replicas[target], set.Spec.MinReadySeconds) || isTerminating(replicas[target]) {
```

Comment:
Why we need to consider `isTerminating`?

Reply:
Terminating is not a phase (such as Pending, Running, Succeeded, Failed, Unknown). Under the hoods, `isTerminating` just checks if the pod DeletionTimestamp is not nil. Current `isHealthy` implementation (`isRunningAndReady(pod) && !isTerminating(pod)`) is quite similar to the line that substituted `isHealthy`. But, there isn't a method (like `isHealthy`) that encapsulates that logic for `isRunningAndAvailable` yet.

There are some **very very** related bugs (on which I'm working on for the next few days) in "StatefulSetMinReadySeconds" feature. They're related to `isRunningAndAvailable` and `isRunningAndReady` methods. I'll probably focus on them, unblocking KEP-961 along the way. What the `isRunningAndAvailable` method does will probably change a little bit, but that previous commit will likely stay almost the same.

Reply:
I followed the above steps to load the tar file in the kind cluster but getting `unknown flag: --authentication-kubeconfig`. Is it expected?

Reply:
> I followed the above steps to load the tar file in the kind cluster but getting `unknown flag: --authentication-kubeconfig`. Is it expected?

Not expected. This flag (`--authentication-kubeconfig`) is set by the command specified here `spec.containers.command` in the "/etc/kubernetes/manifests/kube-controller-manager.yaml" file that I used. No "unknown flag" error at all.

After finding out the ID of the docker container running Kind, I run `docker cp kube-controller-manager.yaml SUBSTITUTE_WITH_THE_KIND_DOCKER_CONTAINER_ID:/etc/kubernetes/manifests/kube-controller-manager.yaml`. Then, I delete the old kube controller manager pod so it is recreated with the new manifest.

Reply:
Here's the `kube-controller-manager.yaml` I used for reference:

```yaml
apiVersion: v1
kind: Pod
metadata:
  creationTimestamp: null
  labels:
    component: kube-controller-manager
    tier: control-plane
  name: kube-controller-manager
  namespace: kube-system
spec:
  containers:
  - command:
    - kube-controller-manager
    - --allocate-node-cidrs=true
    - --authentication-kubeconfig=/etc/kubernetes/controller-manager.conf
    - --authorization-kubeconfig=/etc/kubernetes/controller-manager.conf
    - --bind-address=127.0.0.1
    - --client-ca-file=/etc/kubernetes/pki/ca.crt
    - --cluster-cidr=10.244.0.0/16
    - --cluster-name=lab
    - --cluster-signing-cert-file=/etc/kubernetes/pki/ca.crt
    - --cluster-signing-key-file=/etc/kubernetes/pki/ca.key
    - --controllers=*,bootstrapsigner,tokencleaner
    - --enable-hostpath-provisioner=true
    - --feature-gates=MaxUnavailableStatefulSet=true
    - --kubeconfig=/etc/kubernetes/controller-manager.conf
    - --leader-elect=true
    - --requestheader-client-ca-file=/etc/kubernetes/pki/front-proxy-ca.crt
    - --root-ca-file=/etc/kubernetes/pki/ca.crt
    - --service-account-private-key-file=/etc/kubernetes/pki/sa.key
    - --service-cluster-ip-range=10.96.0.0/16
    - --use-service-account-credentials=true
    # image: registry.k8s.io/kube-controller-manager:v1.27.3
    image: registry.k8s.io/kube-controller-manager-amd64:v1.30.0-beta.0.13_bfce32798dbaef-dirty
    imagePullPolicy: IfNotPresent
    livenessProbe:
      failureThreshold: 8
      httpGet:
        host: 127.0.0.1
        path: /healthz
        port: 10257
        scheme: HTTPS
      initialDelaySeconds: 10
      periodSeconds: 10
      timeoutSeconds: 15
    name: kube-controller-manager
    resources:
      requests:
        cpu: 200m
    startupProbe:
      failureThreshold: 24
      httpGet:
        host: 127.0.0.1
        path: /healthz
        port: 10257
        scheme: HTTPS
      initialDelaySeconds: 10
      periodSeconds: 10
      timeoutSeconds: 15
    volumeMounts:
    - mountPath: /etc/ssl/certs
      name: ca-certs
      readOnly: true
    - mountPath: /etc/ca-certificates
      name: etc-ca-certificates
      readOnly: true
    - mountPath: /usr/libexec/kubernetes/kubelet-plugins/volume/exec
      name: flexvolume-dir
    - mountPath: /etc/kubernetes/pki
      name: k8s-certs
      readOnly: true
    - mountPath: /etc/kubernetes/controller-manager.conf
      name: kubeconfig
      readOnly: true
    - mountPath: /usr/local/share/ca-certificates
      name: usr-local-share-ca-certificates
      readOnly: true
    - mountPath: /usr/share/ca-certificates
      name: usr-share-ca-certificates
      readOnly: true
  hostNetwork: true
  priority: 2000001000
  priorityClassName: system-node-critical
  securityContext:
    seccompProfile:
      type: RuntimeDefault
  volumes:
  - hostPath:
      path: /etc/ssl/certs
      type: DirectoryOrCreate
    name: ca-certs
  - hostPath:
      path: /etc/ca-certificates
      type: DirectoryOrCreate
    name: etc-ca-certificates
  - hostPath:
      path: /usr/libexec/kubernetes/kubelet-plugins/volume/exec
      type: DirectoryOrCreate
    name: flexvolume-dir
  - hostPath:
      path: /etc/kubernetes/pki
      type: DirectoryOrCreate
    name: k8s-certs
  - hostPath:
      path: /etc/kubernetes/controller-manager.conf
      type: FileOrCreate
    name: kubeconfig
  - hostPath:
      path: /usr/local/share/ca-certificates
      type: DirectoryOrCreate
    name: usr-local-share-ca-certificates
  - hostPath:
      path: /usr/share/ca-certificates
      type: DirectoryOrCreate
    name: usr-share-ca-certificates
status: {}
```

Reply:
> I followed the above steps to load the tar file in the kind cluster but getting `unknown flag: --authentication-kubeconfig`. Is it expected?

Which command returns that error?